src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidZeroLengthArrayAllocations.Fixer.cs

```diff
@@ -46,20 +48,50 @@ public sealed override FixAllProvider GetFixAllProvider()
             SemanticModel semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             SyntaxGenerator generator = editor.Generator;
 
-            SyntaxNode variableDecl = generator.GetDeclaration(nodeToFix, DeclarationKind.Variable);
-            SyntaxNode typeNode = generator.GetType(variableDecl);
+            var operation = semanticModel.GetOperationInternal(nodeToFix, cancellationToken);
+            var arrayCreationExpression = FindArrayCreation(operation);
+            nodeToFix = arrayCreationExpression.Syntax;
 
-            var type = semanticModel.GetTypeInfo(typeNode, cancellationToken).Type as IArrayTypeSymbol;
-            INamedTypeSymbol arrayTypeSymbol = semanticModel.Compilation.GetTypeByMetadataName(AvoidZeroLengthArrayAllocationsAnalyzer.ArrayTypeName);
+            ITypeSymbol elementType = arrayCreationExpression.GetElementType();
+            SyntaxNode arrayEmptyInvocation = GenerateArrayEmptyInvocation(generator, elementType, semanticModel).WithTriviaFrom(nodeToFix);
 
-            SyntaxNode arrayEmptyName = generator.QualifiedName(generator.TypeExpression(arrayTypeSymbol),
-                                                         generator.GenericName(AvoidZeroLengthArrayAllocationsAnalyzer.ArrayEmptyMethodName, type.ElementType));
-            SyntaxNode arrayEmptyInvocation = generator.InvocationExpression(arrayEmptyName);
-            arrayEmptyInvocation = arrayEmptyInvocation.WithLeadingTrivia(nodeToFix.GetLeadingTrivia()).WithTrailingTrivia(nodeToFix.GetTrailingTrivia());
             editor.ReplaceNode(nodeToFix, arrayEmptyInvocation);
             return editor.GetChangedDocument();
         }
 
+        private static IArrayCreationExpression FindArrayCreation(IOperation operation)
+        {
+            switch (operation.Kind)
+            {
+                case OperationKind.Argument:
+                    // Happens if the array is passed as a method argument as it's being created.
+                    // e.g. M(new object[0])
+                    return FindArrayCreation(
+                        ((IArgument)operation).Value);
+
+                case OperationKind.ArrayCreationExpression:
+                    return (IArrayCreationExpression)operation;
+
+                case OperationKind.ConversionExpression:
+                    // Happens when the array is converted to a different type as it's created.
+                    // e.g. M's signature is M(object obj), M(new object[0])
+                    //      In this case, the IConversionExpression is the child of the IArgument.
+                    return FindArrayCreation(
```

Comment:
@333fred Referring to the email you sent, do you mean the changes to `GetOperation` would make this unnecessary? If so should I add a TODO to revisit this?

Reply:
It should still be necessary in the case of an explicit cast as an argument to a function, for example `M((object)new object[10])`. Could you add a test case for that, and also a test case for something like `M((string[])new object[0])`, to make sure we're supplying the correct generic argument.

---
In reply to: [149532885](https://github.com/dotnet/roslyn-analyzers/pull/1339#discussion_r149532885) [](ancestors = 149532885)